beacon-chain/core/validators/validator.go

```diff
@@ -649,40 +650,112 @@ func UpdateRegistry(state *pb.BeaconState) (*pb.BeaconState, error) {
+	for idx, validator := range state.ValidatorRegistry {
 		// Exit validators within the allowable balance churn.
 		if validator.ExitSlot > state.Slot+config.EntryExitDelay &&
 			validator.StatusFlags == pb.ValidatorRecord_INITIATED_EXIT {
-			balChurn += EffectiveBalance(state, uint32(index))
+			balChurn += EffectiveBalance(state, uint32(idx))
 			if balChurn > maxBalChurn {
 				break
 			}
-			state, err = ExitValidator(state, uint32(index))
+			state, err = ExitValidator(state, uint32(idx))
 			if err != nil {
-				return nil, fmt.Errorf("could not exit validator %d: %v", index, err)
+				return nil, fmt.Errorf("could not exit validator %d: %v", idx, err)
 			}
 		}
 	}
 	state.ValidatorRegistryLatestChangeSlot = state.Slot
 	return state, nil
 }
 
+// ProcessPenaltiesAndExits prepares the validators and the penalized validators
+// for withdrawal.
+//
+// Spec pseudocode definition:
+// def process_penalties_and_exits(state: BeaconState) -> None:
+//    # The active validators
+//    active_validator_indices = get_active_validator_indices(state.validator_registry, state.slot)
+//    # The total effective balance of active validators
+//    total_balance = sum([get_effective_balance(state, i) for i in active_validator_indices])
+//
+//    for index, validator in enumerate(state.validator_registry):
+//        if (state.slot // EPOCH_LENGTH) == (validator.penalized_slot // EPOCH_LENGTH) + LATEST_PENALIZED_EXIT_LENGTH // 2:
+//            e = (state.slot // EPOCH_LENGTH) % LATEST_PENALIZED_EXIT_LENGTH
+//            total_at_start = state.latest_penalized_exit_balances[(e + 1) % LATEST_PENALIZED_EXIT_LENGTH]
+//            total_at_end = state.latest_penalized_exit_balances[e]
+//            total_penalties = total_at_end - total_at_start
+//            penalty = get_effective_balance(state, index) * min(total_penalties * 3, total_balance) // total_balance
+//            state.validator_balances[index] -= penalty
+//
+//    all_indices = list(range(len(state.validator_registry)))
+//    eligible_indices = filter(eligible, all_indices)
+//    sorted_indices = sorted(eligible_indices, key=lambda index: state.validator_registry[index].exit_count)
+//    withdrawn_so_far = 0
+//    for index in sorted_indices:
+//        prepare_validator_for_withdrawal(state, index)
+//        withdrawn_so_far += 1
+//        if withdrawn_so_far >= MAX_WITHDRAWALS_PER_EPOCH:
+//            break
+func ProcessPenaltiesAndExits(state *pb.BeaconState) *pb.BeaconState {
+
+	activeValidatorIndices := ActiveValidatorIndices(state.ValidatorRegistry, state.Slot)
+	totalBalance := TotalEffectiveBalance(state, activeValidatorIndices)
+
+	for idx, validator := range state.ValidatorRegistry {
+		penalized := validator.PenalizedSlot/config.EpochLength +
+			config.LatestPenalizedExitLength/2
+		if state.Slot/config.EpochLength == penalized {
+			e := (state.Slot / config.EpochLength) % config.LatestPenalizedExitLength
```

Comment:
can better variable  naming be used than `e` ? I know its from the spec but I have no idea what e is referring to

Reply:
+1, its not clear what `e` is in this context